snapcraft/parts/parts.py

```diff
@@ -0,0 +1,143 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Craft-parts lifecycle wrapper."""
+
+import pathlib
+from typing import Any, Dict
+
+import craft_parts
+from craft_cli import emit
+from craft_parts import ActionType, Step
+from xdg import BaseDirectory  # type: ignore
+
+from snapcraft.errors import PartsLifecycleError
+
+_LIFECYCLE_STEPS = {
+    "pull": Step.PULL,
+    "overlay": Step.OVERLAY,
+    "build": Step.BUILD,
+    "stage": Step.STAGE,
+    "prime": Step.PRIME,
+}
+
+
+class PartsLifecycle:
+    """Create and manage the parts lifecycle.
+
+    :param all_parts: A dictionary containing the parts defined in the project.
+    :param work_dir: The working directory for parts processing.
+
+    :raises PartsLifecycleError: On error initializing the parts lifecycle.
+    """
+
+    def __init__(
+        self, all_parts: Dict[str, Any], *, work_dir: pathlib.Path,
+    ):
+        emit.progress("Initializing parts lifecycle")
+
+        # set the cache dir for parts package management
+        cache_dir = BaseDirectory.save_cache_path("snapcraft")
+
+        try:
+            self._lcm = craft_parts.LifecycleManager(
+                {"parts": all_parts},
+                application_name="snapcraft",
+                work_dir=work_dir,
+                cache_dir=cache_dir,
+                ignore_local_sources=["*.snap"],
+            )
+        except craft_parts.PartsError as err:
+            raise PartsLifecycleError(str(err)) from err
+
+    @property
+    def prime_dir(self) -> pathlib.Path:
+        """Return the parts prime directory path."""
+        return self._lcm.project_info.prime_dir
+
+    def run(self, step_name: str) -> None:
+        """Run the parts lifecycle.
+
+        :param target_step: The final step to execute.
+
+        :raises PartsLifecycleError: On error during lifecycle.
+        :raises RuntimeError: On unexpected error.
+        """
+        target_step = _LIFECYCLE_STEPS.get(step_name)
+        if not target_step:
+            raise RuntimeError(f"Invalid target step {step_name!r}")
```

Comment:
This could of been a KeyError instead, right?

Reply:
Yes, it would generate a KeyError if retrieved directly. I used RuntimeError because the implementation as a dictionary is an internal detail, it could have been implemented as cascading if/else.